File: Python/models/Exemplar.py
import json
from models.Modelo import Modelo

class Exemplar:
    def __init__(self, id, edicao, editora, situacao, idLivro, idGenero):
        self.set_id(id)
        self.set_edicao(edicao)
        self.set_editora(editora)
        self.set_situacao(situacao)
        self.set_id_livro(idLivro)
        self.set_id_genero(idGenero)

    def set_id(self, id):
        self.__id = id
    
    def set_edicao(self, edicao):
        if not edicao:
            raise ValueError("Campo edição não pode ser vazio.")
        if edicao < 1:
            raise ValueError("Edição inválida.")
        self.__edicao = edicao
    
    def set_editora(self, editora):
        if not editora:
            raise ValueError("Campo editora não pode ser vazio.")
        self.__editora = editora

    def set_situacao(self, situacao):
        if situacao == "":
            raise ValueError("Campo situação não pode ser vazio.")
        self.__situacao = situacao

    def set_id_livro(self, idLivro):
        if not idLivro:
            raise ValueError("Campo livro não pode ser vazio.")
        self.__id_livro = idLivro

    def set_id_genero(self, idGenero):
        if not idGenero:
            raise ValueError("Campo gênero não pode ser vazio.")
        self.__id_genero = idGenero

    def get_id(self):
        return self.__id
    
    def get_edicao(self):
        return self.__edicao
    
    def get_editora(self):
        return self.__editora

    def get_situacao(self):
        return self.__situacao

    def get_id_livro(self):
        return self.__id_livro

    def get_id_genero(self):
        return self.__id_genero
    
    def __str__(self):
        from view.view import View
        situacao = "Disponível" if self.get_situacao() else "Emprestado"

        livro = View.livro_listar()
        genero = View.genero_listar()

        nome_livro = next((l.get_titulo() for l in livro if l.get_id() == self.get_id_livro()), "Desconhecido")
        nome_genero = next((g.get_descricao() for g in genero if g.get_id() == self.get_id_genero()), "Desconhecido")
        return f"EDICAO = {self.get_edicao()}  |  EDITORA = {self.get_editora()}  |  SITUACAO = {situacao} | LIVRO = {nome_livro} | GENERO = {nome_genero}"
    
    def to_dict(self):
        return {
            "id": self.get_id(), "edicao": self.get_edicao(), "editora" : self.get_editora(), "situacao" : self.get_situacao(), "idLivro" : self.get_id_livro(), "idGenero" : self.get_id_genero()
        }
class Exemplares(Modelo):
    @classmethod
    def salvar(cls):
        with open("../Arquivos/exemplares.json", mode="w") as arquivo:
            json.dump([exemplar.to_dict() for exemplar in cls.objetos], arquivo, indent=6) #vars - converte um objeto em dicionario
            #dump - pega a lista de obejtos e salva no arquivo
            
    @classmethod
    def abrir(cls):
        cls.objetos = []
        try:
            with open("../Arquivos/exemplares.json", mode="r") as arquivo:
                exemplares_json = json.load(arquivo)
                for obj in exemplares_json:
                    e = Exemplar(obj["id"], obj["edicao"], obj["editora"], obj["situacao"], obj["idLivro"], obj["idGenero"])
                    cls.objetos.append(e)    
        except FileNotFoundError:
            pass

File: Java/src/models/Exemplar.java
public class Exemplar implements Objeto{
    private int id;
    private int edicao;
    private String editora;
    private Boolean situacao;
    private int idLivro;
    private int idGenero;

    public Exemplar(int id, int edicao, String editora, Boolean situacao, int idLivro, int idGenero){
        setId(id);
        setEdicao(edicao);
        setEditora(editora);
        setSituacao(situacao);
        setIdLivro(idLivro);
        setIdGenero(idGenero);
    }

    public void setId(int id){
        this.id = id;
    }

    public void setEdicao(int edicao){
        if (edicao > 0) this.edicao = edicao;
        else throw new IllegalArgumentException("Edição inválida");
    }

    public void setEditora(String editora){
        if (editora != "") this.editora = editora;
        else throw new IllegalArgumentException("Editora inválida");
    }

    public void setSituacao(Boolean situacao){
        this.situacao = situacao;
    }

    public void setIdLivro(int idLivro){
        this.idLivro = idLivro;
    }

    public void setIdGenero(int idGenero){
        this.idGenero = idGenero;
    }

    public int getId(){
        return this.id;
    }

    public int getEdicao(){
        return this.edicao;
    }

    public String getEditora(){
        return this.editora;
    }

    public Boolean getSituacao(){
        return this.situacao;
    }

    public int getIdLivro(){
        return this.idLivro;
    }

    public int getIdGenero(){
        return this.idGenero;
    }

    public String toString(){
        return String.format("Exemplar \n ID = %s \n Edição = %s \n Editora = %s \n Situacao = %s \n Id Livro = %d \n Id genero = %s", this.id, this.edicao, this.editora, this.situacao, this.idLivro, this.idGenero);
    }
}